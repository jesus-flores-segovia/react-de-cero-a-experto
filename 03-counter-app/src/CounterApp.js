import {useState} from "react"
import PropTypes from "prop-types";


const CounterApp = ({value = CounterApp.defaultProps.value}) => {

    // Definimos el State del componente con el valor a 0 mediante el Hook useState
    const [counter, setCounter] = useState(value);

    const handleSubstract = (e) => {
        
        // En caso de no disponer de la desestructuración del array
        //setCounter((value) => value + 1);
        setCounter(counter - 1);
    };

    const handleReset = (e) => {
        
        // En caso de no disponer de la desestructuración del array
        //setCounter((value) => value + 1);
        setCounter(value);
    };
    
    const handleAdd = (e) => {
        
        // En caso de no disponer de la desestructuración del array
        //setCounter((value) => value + 1);
        setCounter(counter + 1);
    };

    return (

        <>
            <h1>Counter App</h1>
            <h2>{counter}</h2>
            <button onClick={handleSubstract}>-1</button>
            <button onClick={handleReset}>Reset</button>
            <button onClick={handleAdd}>+1</button>
        </>
    );

};

CounterApp.propTypes = {
    value: PropTypes.number
};

CounterApp.defaultProps = {
    value: 0
}

export default CounterApp;