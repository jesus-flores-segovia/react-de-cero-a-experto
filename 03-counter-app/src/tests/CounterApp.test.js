import React from "react";
import CounterApp from "../CounterApp";
import {shallow} from "enzyme";
import "@testing-library/jest-dom";

describe("Tests inside file CounterApp.test.js", () =>{

    // That's not a good practice, but if we initialize this, we can acces to the Intellij information
    let wrapper = shallow(<CounterApp></CounterApp>);

    beforeEach(() => {
        wrapper = shallow(<CounterApp></CounterApp>);
    });

    test("'CounterApp' component must show correctly", () =>{

        expect(wrapper).toMatchSnapshot();
    });

    test("'CounterApp' component must contains the counter state configured by the prop value", () =>{

        const counter = 100;
        const wrapper = shallow(<CounterApp value={counter}></CounterApp>);
        console.log(wrapper.html());

        const counterText = wrapper.find("h2").text();

        expect(counterText).toBe("100")
    });

    test("'CounterApp' add button must increment the counter value", () =>{
        
        const addButton = wrapper.find("button").at(2);
        console.log(addButton.html());

        addButton.simulate("click");

        const counterText = wrapper.find("h2").text();
        expect(counterText).toBe("1")
    });

    test("'CounterApp' substract button must decrease the counter value", () =>{

        const substractButton = wrapper.find("button").at(0);
        console.log(substractButton.html());

        substractButton.simulate("click");

        const counterText = wrapper.find("h2").text();
        expect(counterText).toBe("-1")      
    });

    test("'CounterApp' reset button must re-configure the counter value to 0", () =>{

        const counter = 100;
        const wrapper = shallow(<CounterApp value={counter}></CounterApp>);

        const resetButton = wrapper.find("button").at(1);
        console.log(resetButton.html());

        resetButton.simulate("click");

        const counterText = wrapper.find("h2").text();
        expect(counterText).toBe("100")      
    });
});