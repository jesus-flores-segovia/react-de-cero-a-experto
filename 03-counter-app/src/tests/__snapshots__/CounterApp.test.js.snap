// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file CounterApp.test.js 'CounterApp' component must show correctly 1`] = `
<Fragment>
  <h1>
    Counter App
  </h1>
  <h2>
    0
  </h2>
  <button
    onClick={[Function]}
  >
    -1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</Fragment>
`;
